### PR TITLE
Timeline Features

### PR DESCRIPTION
**Pull request by wubojin**
What was done:

1. No calendar as it is more GUI based than CLI, **auto sort Events** by remaining time left is good enough.
a. Sort by `Ongoing`, `XX minutes left`, `XX hours left`, `XX days left`, `Completed` when apps open
b. However, when an event becomes Completed, it does not get re-sorted to the bottom while the app is open.
KIV: has a potential solution but fail tests

2. UI: **Display current datetime** on the right of the NavBar
3. UI: **Show XX days left** on the top right of EventCard
4. Command: **upcoming events in the next N days** eg. `upcoming 2`
5. Command: **upcoming events on yyyy-MM-dd day** eg. `upcoming 2024-12-10`

What needs to be done in future PR:
- fix auto sort events
- fix upcoming past events

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -56,6 +56,9 @@ dependencies {
     implementation group: 'org.openjfx', name: 'javafx-graphics', version: javaFxVersion, classifier: 'win'
     implementation group: 'org.openjfx', name: 'javafx-graphics', version: javaFxVersion, classifier: 'mac'
     implementation group: 'org.openjfx', name: 'javafx-graphics', version: javaFxVersion, classifier: 'linux'
+    // implementation group: 'org.openjfx', name: 'javafx-swing', version: javaFxVersion, classifier: 'win'
+    // implementation group: 'org.openjfx', name: 'javafx-swing', version: javaFxVersion, classifier: 'mac'
+    // implementation group: 'org.openjfx', name: 'javafx-swing', version: javaFxVersion, classifier: 'linux'
```
Comment (FriedCabbageSalad) [inline build.gradle:61]: Are these comments necessary to leave in?

File: src/main/java/seedu/address/model/types/common/EventUpcomingPredicate.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+     */
+    public EventUpcomingPredicate(Integer range) {
+        if (range >= 0) {
+            startDate = DateTimeUtil.getCurrentDateTime();
+            endDate = DateTimeUtil.getCurrentDateTime().withHour(23).withMinute(59).withSecond(59).withNano(999999999)
+                    .plusDays(range);
+        } else {
+            startDate = DateTimeUtil.getCurrentDateTime().withHour(0).withMinute(0).withSecond(0).withNano(0)
+                    .minusDays(-range);
+            endDate = DateTimeUtil.getCurrentDateTime();
```
Comment (FriedCabbageSalad) [inline src/main/java/seedu/address/model/types/common/EventUpcomingPredicate.java:29]: Can we use LocalDateTime's now() instead?

File: src/main/java/seedu/address/model/types/common/EventUpcomingPredicate.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        if (other == this) {
+            return true;
+        }
+
+        // instanceof handles nulls
+        if (!(other instanceof EventNameContainsKeywordsPredicate)) {
```
Comment (FriedCabbageSalad) [inline src/main/java/seedu/address/model/types/common/EventUpcomingPredicate.java:56]: Shouldn't "EventNameContainsKeywordsPredicate" be EventUpcomingPredicate?

**Review by FriedCabbageSalad — COMMENTED** [review]
Other than the naming error, rest are stylistic. LGTM otherwise!

File: src/main/java/seedu/address/model/types/common/EventUpcomingPredicate.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+    /**
+     * Constructs an EventUpcomingPredicate with a positive or negative number of days
+     * This constructor is used to create a filter for the past N or next N days of events
+     * @param range the number of days in the future/past.
+     */
+    public EventUpcomingPredicate(Integer range) {
+        if (range >= 0) {
+            startDate = DateTimeUtil.getCurrentDateTime();
+            endDate = DateTimeUtil.getCurrentDateTime().withHour(23).withMinute(59).withSecond(59).withNano(999999999)
```
Comment (FriedCabbageSalad) [inline src/main/java/seedu/address/model/types/common/EventUpcomingPredicate.java:24]: Is the Seconds/Nano seconds necessary?

File: src/main/java/seedu/address/model/types/common/EventUpcomingPredicate.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    /**
+     * Constructs an UpcomingEventCommand with a specific date
+     * This constructor is used to create a filter for events that happen on that date
+     * @param date the specific date.
+     */
+    public EventUpcomingPredicate(DateTime date) {
+        startDate = date.toLocalDateTime().withHour(0).withMinute(0).withSecond(0).withNano(0);
+        endDate = date.toLocalDateTime().withHour(23).withMinute(59).withSecond(59).withNano(999999999);
```
Comment (FriedCabbageSalad) [inline src/main/java/seedu/address/model/types/common/EventUpcomingPredicate.java:40]: Same here for the Seconds and Nano Seconds

File: src/main/java/seedu/address/model/types/event/UniqueEventList.java [hunk 6]
```diff
@@ -28,6 +31,50 @@ public class UniqueEventList implements Iterable<Event> {
...
             FXCollections.unmodifiableObservableList(internalList);
 
+    /**
+     * Periodically sorts the events especially when an event becomes completed.
+     */
+    //    public UniqueEventList() {
+    //        Timeline eventReSortTimeline = DateTimeUtil.createTimeline(this::sortEvents,
+    //                javafx.util.Duration.minutes(1));
+    //        eventReSortTimeline.play();
+    //    }
```
Comment (FriedCabbageSalad) [inline src/main/java/seedu/address/model/types/event/UniqueEventList.java:41]: Large commented unused segments not good for code quality

**Review by FriedCabbageSalad — COMMENTED** [review]
Some more additional reviews that aren't pressing

File: build.gradle (hunk 1 above)
Comment (Jayden-Kim-NUS) [reply build.gradle:61]: They are commented out to fix a set of failing test cases, see #88

File: src/main/java/seedu/address/model/types/common/DateTime.java [hunk 7]
```diff
@@ -18,7 +23,10 @@ public class DateTime {
     public static final String VALIDATION_REGEX = "\\d{4}-(0[1-9]|1[0-2])-(0[1-9]|[12][0-9]|3[01])"
         + " ([01][0-9]|2[0-3]):[0-5][0-9]";
 
+    private static final DateTimeFormatter DATE_TIME_FORMATTER = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm");
```
Comment (antgyf) [inline src/main/java/seedu/address/model/types/common/DateTime.java:26]: do we eventually want to allow for other separators, eg. yyyy/MM/dd

File: src/main/java/seedu/address/model/types/event/UniqueEventList.java [hunk 8]
```diff
@@ -28,6 +31,50 @@ public class UniqueEventList implements Iterable<Event> {
     private final ObservableList<Event> internalUnmodifiableList =
             FXCollections.unmodifiableObservableList(internalList);
 
+    /**
```
Comment (antgyf) [inline src/main/java/seedu/address/model/types/event/UniqueEventList.java:34]: Same comments as Kieran, let's try not to have too much code commented out

File: src/main/java/seedu/address/model/types/event/UniqueEventList.java [hunk 9]
```diff
@@ -28,6 +31,50 @@ public class UniqueEventList implements Iterable<Event> {
...
+     */
+    //    public UniqueEventList() {
+    //        Timeline eventReSortTimeline = DateTimeUtil.createTimeline(this::sortEvents,
+    //                javafx.util.Duration.minutes(1));
+    //        eventReSortTimeline.play();
+    //    }
+
+    private void sortEvents() {
+        internalList.sort((event1, event2) -> {
+            long remainingTime1 = getEventTimeRemaining(event1);
```
Comment (antgyf) [inline src/main/java/seedu/address/model/types/event/UniqueEventList.java:45]: Nitpicking, but maybe can rename these variables to something clearer, eg. e1RemainingTime and e2RemainingTime

**Pull request merged**